File: frontend/src/services/news.ts
import { apiFetch } from './api-client'

// ─── Types ─────────────────────────────────────────────

export interface NewsArticle {
  id: string
  title: string
  summary: string | null
  content?: string
  url: string | null
  category: string | null
  language: string | null
  published_at: string | null
  collected_at: string | null
  status: string | null
  source_name: string | null
  importance: string | null
  entities?: NewsEntity[]
  events?: NewsEvent[]
}

export interface NewsEntity {
  id: string
  name: string
  entity_type: string
  description: string | null
  confidence: number | null
}

export interface NewsEvent {
  id: string
  event_type: string
  title: string
  summary: string | null
  event_time: string | null
  impact_score: number | null
  impact_direction: string | null
  market?: string | null
  sector?: string | null
  confidence?: number | null
  created_at?: string | null
  article_title?: string | null
}

export interface NewsSource {
  id: string
  source_id: string
  name: string
  source_type: string
  category: string | null
  market: string | null
  priority: number | null
  enabled: boolean
  created_at: string | null
}

export interface SourceHealth {
  source_id: string
  name: string
  source_type: string
  category: string[] | null
  market: string[] | null
  priority: string | null
  enabled: boolean
  last_latency_ms: number | null
  last_success: boolean | null
  last_error: string | null
  error_count: number | null
  last_collected_at: string | null
  total_runs: number
  success_count: number
  failed_count: number
  avg_latency_ms: number
  total_articles: number
  total_stored: number
  total_duplicates: number
  status: 'disabled' | 'no_data' | 'healthy' | 'degraded' | 'error'
}

export interface SourceHealthResponse {
  sources: SourceHealth[]
  coverage: number
  total_sources: number
  enabled_sources: number
  healthy_sources: number
  days: number
}

export interface NewsListResponse {
  articles: NewsArticle[]
  total: number
  limit: number
  offset: number
}

export interface NewsFeedArticle {
  id: string
  title: string
  summary: string | null
  url: string | null
  category: string | null
  language: string | null
  importance_score: number | null
  tier: number | null
  published_at: string | null
  collected_at: string | null
  status: string | null
  source_name: string | null
}

export interface NewsFeedHotTopic {
  name: string
  mentions: number
}

export interface NewsFeed {
  hours: number
  breaking: NewsFeedArticle[]
  high_impact: NewsFeedArticle[]
  hot_topics: NewsFeedHotTopic[]
  summary: {
    breaking: number
    high_impact: number
    hot_topics: number
  }
}

export type QueueState = 'waiting' | 'processing' | 'published' | 'failed'

export interface NewsQueueItem {
  article_id: string
  title: string | null
  source_name: string | null
  published_at: string | null
  importance_score: number | null
  package_id: string | null
  package_status: string | null
  retry_count: number
  state: QueueState
  error: string | null
}

export interface NewsIntelligenceQueue {
  days: number
  summary: Record<QueueState, number>
  items: NewsQueueItem[]
  total: number
}

export interface EventListResponse {
  events: NewsEvent[]
  total: number
}

export interface EventImpact {
  id: string
  event_type: string
  title: string
  summary: string | null
  event_time: string | null
  impact_score: number | null
  impact_direction: string | null
  market: string | null
  sector: string | null
  confidence: number | null
  article_title: string | null
  article_category: string | null
}

export interface CoreEvent {
  id: string
  event_type: string
  title: string
  description: string | null
  event_date: string | null
  entities: string[]
  company_count: number
  impact: { score?: number; direction?: string; duration?: string } | null
  confidence: number | null
  created_at: string | null
}

export interface EventMonitorResponse {
  source: string
  days: number
  today_new: number
  window_total: number
  avg_score: number | null
  direction: { positive: number; negative: number; neutral: number }
  affected_companies: string[]
  company_mentions: Array<{ company: string; event_count: number }>
  events: CoreEvent[]
  total: number
}

export interface TopImpactEvent {
  id: string
  event_type: string
  title: string
  description: string | null
  event_date: string | null
  companies: string[]
  company_count: number
  impact: { score?: number; direction?: string; duration?: string } | null
  score: number | null
  stars: number
  confidence: number | null
}

export interface TopImpactResponse {
  source: string
  days: number
  items: TopImpactEvent[]
  total: number
}

export interface EventTimelineResponse {
  source: string
  entity_name: string
  days: number
  items: TopImpactEvent[]
  total: number
}

export interface ImpactAnalysis {
  entity_name: string
  days?: number
  total_events: number
  positive_count: number
  negative_count: number
  neutral_count: number
  avg_impact_score: number
  events: Array<{
    entity_name: string
    entity_type: string | null
    event_type: string | null
    event_title: string | null
    impact_score: number | null
    impact_direction: string | null
    event_time: string | null
    article_title: string | null
    published_at: string | null
  }>
  message?: string
}

export interface TimelineItem {
  title: string
  category: string | null
  published_at: string | null
  url: string | null
  importance: string | null
  source_name: string | null
}

export interface TimelineResponse {
  entity_name: string
  items: TimelineItem[]
  total: number
}

// ─── API Functions ─────────────────────────────────────

export function fetchNewsArticles(params?: {
  keyword?: string
  category?: string
  days?: number
  limit?: number
  offset?: number
}): Promise<NewsListResponse> {
  const searchParams = new URLSearchParams()
  if (params?.keyword) searchParams.set('keyword', params.keyword)
  if (params?.category) searchParams.set('category', params.category)
  if (params?.days) searchParams.set('days', String(params.days))
  if (params?.limit) searchParams.set('limit', String(params.limit))
  if (params?.offset) searchParams.set('offset', String(params.offset))
  const qs = searchParams.toString()
  return apiFetch<NewsListResponse>(`/api/news/articles${qs ? `?${qs}` : ''}`)
}

export function fetchNewsFeed(params?: {
  hours?: number
  limit?: number
}): Promise<NewsFeed> {
  const searchParams = new URLSearchParams()
  if (params?.hours) searchParams.set('hours', String(params.hours))
  if (params?.limit) searchParams.set('limit', String(params.limit))
  const qs = searchParams.toString()
  return apiFetch<NewsFeed>(`/api/news/feed${qs ? `?${qs}` : ''}`)
}

export function fetchIntelligenceQueue(params?: {
  days?: number
  limit?: number
  state?: QueueState | ''
}): Promise<NewsIntelligenceQueue> {
  const searchParams = new URLSearchParams()
  if (params?.days) searchParams.set('days', String(params.days))
  if (params?.limit) searchParams.set('limit', String(params.limit))
  if (params?.state) searchParams.set('state', params.state)
  const qs = searchParams.toString()
  return apiFetch<NewsIntelligenceQueue>(
    `/api/news/intelligence-queue${qs ? `?${qs}` : ''}`
  )
}

/** 重投失败 Package（failed → draft，复用 DP-D1 Re-Publish 端点） */
export function retryAgentPackage(
  packageId: string
): Promise<{ status: string; message: string; package_id: string }> {
  return apiFetch(`/api/knowledge/packages/${packageId}/retry`, {
    method: 'POST',
  })
}

export function fetchNewsArticle(id: string): Promise<NewsArticle> {
  return apiFetch<NewsArticle>(`/api/news/articles/${id}`)
}

export function fetchNewsEvents(params?: {
  event_type?: string
  entity_name?: string
  days?: number
  limit?: number
}): Promise<EventListResponse> {
  const searchParams = new URLSearchParams()
  if (params?.event_type) searchParams.set('event_type', params.event_type)
  if (params?.entity_name) searchParams.set('entity_name', params.entity_name)
  if (params?.days) searchParams.set('days', String(params.days))
  if (params?.limit) searchParams.set('limit', String(params.limit))
  const qs = searchParams.toString()
  return apiFetch<EventListResponse>(`/api/news/events${qs ? `?${qs}` : ''}`)
}

export function fetchEventImpact(id: string): Promise<EventImpact> {
  return apiFetch<EventImpact>(`/api/news/events/${id}/impact`)
}

/** NIC-B1 Event Monitor：读 core.events（KOC 侧聚合端点） */
export function fetchEventMonitor(params?: {
  event_type?: string
  company?: string
  days?: number
  limit?: number
}): Promise<EventMonitorResponse> {
  const searchParams = new URLSearchParams()
  if (params?.event_type) searchParams.set('event_type', params.event_type)
  if (params?.company) searchParams.set('company', params.company)
  if (params?.days) searchParams.set('days', String(params.days))
  if (params?.limit) searchParams.set('limit', String(params.limit))
  const qs = searchParams.toString()
  return apiFetch<EventMonitorResponse>(
    `/api/knowledge/events/monitor${qs ? `?${qs}` : ''}`
  )
}

/** NIC-B2 Top Impact Events：KOC 分析结果（core.events），星级=影响评分 */
export function fetchTopImpactEvents(params?: {
  company?: string
  days?: number
  limit?: number
}): Promise<TopImpactResponse> {
  const searchParams = new URLSearchParams()
  if (params?.company) searchParams.set('company', params.company)
  if (params?.days) searchParams.set('days', String(params.days))
  if (params?.limit) searchParams.set('limit', String(params.limit))
  const qs = searchParams.toString()
  return apiFetch<TopImpactResponse>(
    `/api/knowledge/events/top-impact${qs ? `?${qs}` : ''}`
  )
}

/** NIC-B3 事件时间线：读 core.events（Timeline 与 Event Monitor 同源） */
export function fetchKnowledgeEventsTimeline(params?: {
  entity_name?: string
  days?: number
  limit?: number
}): Promise<EventTimelineResponse> {
  const searchParams = new URLSearchParams()
  if (params?.entity_name) searchParams.set('entity_name', params.entity_name)
  if (params?.days) searchParams.set('days', String(params.days))
  if (params?.limit) searchParams.set('limit', String(params.limit))
  const qs = searchParams.toString()
  return apiFetch<EventTimelineResponse>(
    `/api/knowledge/events/timeline${qs ? `?${qs}` : ''}`
  )
}

export function fetchNewsImpact(
  entityName: string,
  days?: number
): Promise<ImpactAnalysis> {
  const searchParams = new URLSearchParams({ entity_name: entityName })
  if (days) searchParams.set('days', String(days))
  return apiFetch<ImpactAnalysis>(`/api/news/impact?${searchParams.toString()}`)
}

export function fetchNewsTimeline(
  entityName: string,
  days?: number,
  limit?: number
): Promise<TimelineResponse> {
  const searchParams = new URLSearchParams({ entity_name: entityName })
  if (days) searchParams.set('days', String(days))
  if (limit) searchParams.set('limit', String(limit))
  return apiFetch<TimelineResponse>(`/api/news/timeline?${searchParams.toString()}`)
}

export function fetchNewsSources(enabledOnly?: boolean): Promise<{
  sources: NewsSource[]
  total: number
}> {
  const searchParams = new URLSearchParams()
  if (enabledOnly !== undefined)
    searchParams.set('enabled_only', String(enabledOnly))
  const qs = searchParams.toString()
  return apiFetch<{ sources: NewsSource[]; total: number }>(
    `/api/news/sources${qs ? `?${qs}` : ''}`
  )
}

export function fetchSourceHealth(days?: number): Promise<SourceHealthResponse> {
  const searchParams = new URLSearchParams()
  if (days) searchParams.set('days', String(days))
  const qs = searchParams.toString()
  return apiFetch<SourceHealthResponse>(
    `/api/news/sources/health${qs ? `?${qs}` : ''}`
  )
}

export function setNewsSourceEnabled(
  sourceId: string,
  enabled: boolean
): Promise<{ source_id: string; enabled: boolean; status: string; message: string }> {
  return apiFetch(`/api/news/sources/${encodeURIComponent(sourceId)}/enabled`, {
    method: 'POST',
    body: JSON.stringify({ enabled }),
  })
}

export interface CollectResponse {
  status: string
  message: string
}

export function triggerNewsCollect(keyword: string): Promise<CollectResponse> {
  return apiFetch<CollectResponse>('/api/news/collect', {
    method: 'POST',
    body: JSON.stringify({ keyword, priority: 'high' }),
  })
}

File: langgraph/api/news.py
"""News API Router - 新闻智能管线 REST 端点

提供新闻搜索、文章详情、事件浏览、影响分析、时间线等接口。
数据来源: PostgreSQL news schema（通过 services/news_storage.py）。
"""

import asyncio
import json
import logging
import uuid

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel, Field

from services.news_storage import news_storage
from tools.postgres import postgres_tool

logger = logging.getLogger(__name__)

news_router = APIRouter(prefix="/api/news", tags=["news"])


def _serialize(row: dict) -> dict:
    """序列化 UUID/时间字段为字符串"""
    result = {}
    for k, v in row.items():
        if hasattr(v, "isoformat"):
            result[k] = v.isoformat()
        elif isinstance(v, uuid.UUID):  # UUID
            result[k] = str(v)
        else:
            result[k] = v
    return result


# ──────────────────────────────────────────────
# Articles
# ──────────────────────────────────────────────

@news_router.get("/articles")
async def list_articles(
    keyword: str = Query("", description="搜索关键词"),
    category: str = Query("", description="分类过滤"),
    days: int = Query(7, ge=1, le=365, description="时间范围（天）"),
    limit: int = Query(20, ge=1, le=100, description="每页数量"),
    offset: int = Query(0, ge=0, description="偏移量"),
):
    """搜索新闻文章列表"""
    articles = await news_storage.search_articles(
        keyword=keyword, category=category, days=days, limit=limit, offset=offset,
    )
    total = await news_storage.count_articles(
        keyword=keyword, category=category, days=days,
    )
    return {
        "articles": [_serialize(a) for a in articles],
        "total": total,
        "limit": limit,
        "offset": offset,
    }


@news_router.get("/articles/{article_id}")
async def get_article(article_id: str):
    """获取文章详情（含实体/事件）"""
    article = await news_storage.get_article_by_id(article_id)
    if not article:
        raise HTTPException(status_code=404, detail=f"Article '{article_id}' not found")
    serialized = _serialize(article)
    if "entities" in serialized:
        serialized["entities"] = [_serialize(e) for e in serialized["entities"]]
    if "events" in serialized:
        serialized["events"] = [_serialize(e) for e in serialized["events"]]
    return serialized


@news_router.get("/feed")
async def live_feed(
    hours: int = Query(24, ge=1, le=168, description="Breaking 时间窗口（小时）"),
    limit: int = Query(10, ge=1, le=50, description="每区条数"),
):
    """NIC-A1 Live Feed 分层视图

    - breaking：importance_score >= 0.8 且近 hours 小时发布（高影响 + 新鲜）
    - high_impact：importance_score >= 0.8 或 tier=1（分类器高重要度 / 永久价值）
    - hot_topics：近 hours 小时被提及文章数 Top N（热点实体）
    各查询独立 try/except，单故障不影响整体返回（对应区降级为空）。
    """
    result: dict = {"hours": hours, "breaking": [], "high_impact": [], "hot_topics": []}

    _ARTICLE_SELECT = (
        "SELECT a.id, a.title, a.summary, a.url, a.category, a.language, "
        "a.importance_score, a.tier, a.published_at, a.collected_at, a.status, "
        "s.name AS source_name "
        "FROM news.articles a LEFT JOIN news.sources s ON s.id = a.source_id "
    )

    # ── Breaking：高影响且新鲜（近 hours 小时） ──
    try:
        rows = await postgres_tool.query(
            _ARTICLE_SELECT
            + "WHERE a.importance_score >= 0.8 "
            "AND a.published_at >= NOW() - INTERVAL '1 hour' * $1 "
            "ORDER BY a.importance_score DESC, a.published_at DESC LIMIT $2",
            hours,
            limit,
        )
        result["breaking"] = [_serialize(r) for r in rows]
    except Exception:
        logger.warning("feed: breaking query failed, degraded")

    # ── 高影响：classifier importance 高 或 tier=1（永久价值） ──
    try:
        rows = await postgres_tool.query(
            _ARTICLE_SELECT
            + "WHERE a.importance_score >= 0.8 OR a.tier = 1 "
            "ORDER BY a.importance_score DESC, a.published_at DESC LIMIT $1",
            limit,
        )
        result["high_impact"] = [_serialize(r) for r in rows]
    except Exception:
        logger.exception("feed: high_impact query failed, degraded")
        result["high_impact"] = []

    # ── 热点实体：近 hours 小时实体提及文章数 Top N ──
    try:
        rows = await postgres_tool.query(
            "SELECT e.name, COUNT(DISTINCT e.article_id) AS mentions "
            "FROM news.entities e "
            "JOIN news.articles a ON a.id = e.article_id "
            "WHERE a.published_at >= NOW() - INTERVAL '1 hour' * $1 "
            "GROUP BY e.name ORDER BY mentions DESC LIMIT $2",
            hours,
            limit,
        )
        result["hot_topics"] = [
            {"name": r["name"], "mentions": int(r["mentions"] or 0)} for r in rows
        ]
    except Exception:
        logger.warning("feed: hot_topics query failed, degraded")

    result["summary"] = {
        "breaking": len(result["breaking"]),
        "high_impact": len(result["high_impact"]),
        "hot_topics": len(result["hot_topics"]),
    }
    return result


# ──────────────────────────────────────────────
# Intelligence Queue（NIC-A2）
# ──────────────────────────────────────────────

_IQ_JOIN = (
    "FROM news.articles a "
    "LEFT JOIN news.sources s ON s.id = a.source_id "
    "LEFT JOIN knowledge_packages kp "
    "  ON kp.source_type = 'news' "
    "  AND (kp.document_id = a.id OR kp.payload->'source'->>'url' = a.url) "
)


def _iq_state(package_status: str | None) -> str:
    """Package 状态 → 四态"""
    if package_status is None:
        return "waiting"
    if package_status == "draft":
        return "processing"
    if package_status in ("published", "consumed"):
        return "published"
    return "failed"


def _iq_error(processing_metadata) -> str | None:
    """从 processing_metadata 提取 Package 失败原因（jsonb 可能是 str）"""
    if not processing_metadata:
        return None
    meta = processing_metadata
    if isinstance(meta, str):
        try:
            meta = json.loads(meta)
        except (ValueError, TypeError):
            return str(processing_metadata)[:300]
    publish = meta.get("publish") or {}
    if isinstance(publish, dict):
        return publish.get("last_error")
    return None


@news_router.get("/intelligence-queue")
async def intelligence_queue(
    days: int = Query(7, ge=1, le=90, description="统计时间范围（天）"),
    limit: int = Query(20, ge=1, le=100, description="明细条数"),
    state: str = Query("", description="按四态过滤: waiting/processing/published/failed"),
):
    """NIC-A2 Intelligence Queue 四态联查

    联查 news.articles → knowledge_packages（source_type='NEWS'）：
      - waiting：已采集但无对应 Package（未进入发布链路）
      - processing：Package 已生成但 status=draft（处理/待发布中）
      - published：Package 已发布或已消费（published/consumed）
      - failed：Package 发布/消费失败（含失败原因，可经 retry 重投）

    关联键：kp.document_id = a.id（DP-D2 升级后），
    兼容存量 Package（document_id 为 NULL）按 payload.source.url 匹配。
    各查询独立 try/except，单故障不影响整体返回（对应字段降级）。
    """
    # 防御：直接调用（非 FastAPI 注入）时 Query 默认值是对象而非字符串
    state = state if isinstance(state, str) else ""

    result: dict = {
        "days": days,
        "summary": {"waiting": 0, "processing": 0, "published": 0, "failed": 0},
        "items": [],
        "total": 0,
    }

    # ── 四态计数（单查询聚合，避免 JOIN 重复行膨胀） ──
    try:
        rows = await postgres_tool.query(
            "SELECT "
            "COUNT(*) FILTER (WHERE kp.id IS NULL) AS waiting, "
            "COUNT(*) FILTER (WHERE kp.status = 'draft') AS processing, "
            "COUNT(*) FILTER (WHERE kp.status IN ('published', 'consumed')) AS published, "
            "COUNT(*) FILTER (WHERE kp.status = 'failed') AS failed "
            + _IQ_JOIN
            + "WHERE a.published_at >= NOW() - INTERVAL '1 day' * $1",
            days,
        )
        if rows:
            for k in ("waiting", "processing", "published", "failed"):
                result["summary"][k] = int(rows[0].get(k) or 0)
    except Exception:
        logger.warning("intelligence-queue: summary query failed, degraded")

    # ── 明细（含四态派生与失败原因） ──
    try:
        rows = await postgres_tool.query(
            "SELECT a.id AS article_id, a.title, a.published_at, a.importance_score, "
            "s.name AS source_name, "
            "kp.id AS package_id, kp.status AS package_status, kp.retry_count, "
            "kp.processing_metadata "
            + _IQ_JOIN
            + "WHERE a.published_at >= NOW() - INTERVAL '1 day' * $1 "
            "ORDER BY a.published_at DESC LIMIT $2",
            days,
            limit,
        )
        items = []
        for r in rows:
            st = _iq_state(r.get("package_status"))
            if state and st != state:
                continue
            pub = r.get("published_at")
            items.append({
                "article_id": str(r["article_id"]),
                "title": r.get("title"),
                "source_name": r.get("source_name"),
                "published_at": pub.isoformat() if hasattr(pub, "isoformat") else pub,
                "importance_score": r.get("importance_score"),
                "package_id": str(r["package_id"]) if r.get("package_id") else None,
                "package_status": r.get("package_status"),
                "retry_count": r.get("retry_count") or 0,
                "state": st,
                "error": _iq_error(r.get("processing_metadata")) if st == "failed" else None,
            })
        result["items"] = items
        result["total"] = len(items)
    except Exception:
        logger.exception("intelligence-queue: items query failed, degraded")

    return result


# ──────────────────────────────────────────────
# Events
# ──────────────────────────────────────────────

@news_router.get("/events")
async def list_events(
    event_type: str = Query("", description="事件类型过滤"),
    entity_name: str = Query("", description="关联实体名称"),
    days: int = Query(30, ge=1, le=365, description="时间范围（天）"),
    limit: int = Query(20, ge=1, le=100, description="返回数量"),
):
    """搜索新闻事件"""
    events = await news_storage.search_events(
        event_type=event_type, entity_name=entity_name, days=days, limit=limit,
    )
    return {
        "events": [_serialize(e) for e in events],
        "total": len(events),
    }


@news_router.get("/events/{event_id}/impact")
async def get_event_impact(event_id: str):
    """获取事件影响评估"""
    result = await news_storage.get_event_impact(event_id)
    if not result:
        raise HTTPException(status_code=404, detail=f"Event '{event_id}' not found")
    return _serialize(result)


# ──────────────────────────────────────────────
# Analysis
# ──────────────────────────────────────────────

@news_router.get("/impact")
async def analyze_impact(
    entity_name: str = Query(..., min_length=1, description="实体名称"),
    days: int = Query(30, ge=1, le=365, description="分析时间范围（天）"),
):
    """聚合分析实体近期新闻影响

    DEPRECATED（NIC-B2）：Impact Monitor 已改为消费 KOC 分析结果
    （/api/knowledge/events/top-impact | core.events），不再触发实时重算。
    本端点保留供过渡期兼容。
    """
    results = await news_storage.get_entity_news_impact(entity_name, days=days)

    if not results:
        return {
            "entity_name": entity_name,
            "total_events": 0,
            "positive_count": 0,
            "negative_count": 0,
            "neutral_count": 0,
            "avg_impact_score": 0.0,
            "events": [],
            "message": f"No news found for '{entity_name}' in last {days} days",
        }

    positive = sum(1 for r in results if r.get("impact_direction") == "positive")
    negative = sum(1 for r in results if r.get("impact_direction") == "negative")
    scores = [r["impact_score"] for r in results if r.get("impact_score") is not None]
    avg_score = sum(scores) / len(scores) if scores else 0.0

    return {
        "entity_name": entity_name,
        "days": days,
        "total_events": len(results),
        "positive_count": positive,
        "negative_count": negative,
        "neutral_count": len(results) - positive - negative,
        "avg_impact_score": round(avg_score, 3),
        "events": [_serialize(r) for r in results[:20]],
    }


@news_router.get("/timeline")
async def get_timeline(
    entity_name: str = Query(..., min_length=1, description="实体名称"),
    days: int = Query(90, ge=1, le=365, description="时间范围（天）"),
    limit: int = Query(50, ge=1, le=200, description="返回数量"),
):
    """获取实体新闻时间线"""
    results = await news_storage.get_entity_news_timeline(
        entity_name, days=days, limit=limit,
    )
    return {
        "entity_name": entity_name,
        "items": [_serialize(r) for r in results],
        "total": len(results),
    }


# ──────────────────────────────────────────────
# Collect (Manual Trigger)
# ──────────────────────────────────────────────

class CollectRequest(BaseModel):
    keyword: str = Field("", description="搜索/采集关键词（可选，用于过滤采集结果）")
    priority: str = Field("high", description="采集优先级: high/normal/low")


async def _run_collection(keyword: str, priority: str) -> None:
    """后台执行新闻采集任务（不阻塞请求）"""
    from collectors.source_registry import source_registry
    from collectors.rss_collector import collect_rss
    from collectors.web_collector import collect_web
    from collectors.health_metrics import record_collect_run
    from graphs.news_analysis_graph import get_news_graph
    from monitoring.agent_center import invoke_tracked
    import time

    logger.info("[Collect] Manual trigger | keyword=%r | priority=%s", keyword, priority)
    try:
        sources = source_registry.get_enabled(priority=priority)
        if not sources:
            # 回退：尝试所有已启用源
            sources = source_registry.get_enabled()
        if not sources:
            logger.warning("[Collect] No enabled sources found")
            return

        graph = get_news_graph()
        keyword_lower = keyword.strip().lower()

        for source in sources:
            collect_start = time.monotonic()
            try:
                if source.source_type == "rss":
                    articles = await collect_rss(source)
                elif source.source_type == "crawler":
                    articles = await collect_web(source)
                else:
                    continue

                if not articles:
                    # 采集成功但无文章，仍记录指标
                    await record_collect_run(
                        source.id, source.name, True,
                        latency_ms=int((time.monotonic() - collect_start) * 1000),
                        articles_fetched=0, articles_stored=0,
                    )
                    continue

                # 关键词过滤（若提供）
                if keyword_lower:
                    articles = [
                        a for a in articles
                        if keyword_lower in (a.get("title", "") + " " + a.get("content", "")).lower()
                    ]
                    if not articles:
                        logger.info("[Collect] No matching articles for keyword in source=%s", source.id)
                        continue

                result = await invoke_tracked(
                    graph,
                    {
                        "source_id": source.id,
                        "raw_articles": articles,
                        "cleaned_articles": [],
                        "unique_articles": [],
                        "classified_articles": [],
                        "entities": [],
                        "events": [],
                        "relations": [],
                        "impact_assessments": [],
                        "stored_article_ids": [],
                        "stored_event_ids": [],
                        "knowledge_agent_triggered": False,
                        "errors": [],
                    },
                    agent_id="news_intelligence",
                    question=f"manual collect source={source.id} keyword={keyword}",
                )

                stored = len(result.get("stored_article_ids", []))
                # 采集健康指标（Latency/Articles/Duplicates）
                await record_collect_run(
                    source.id, source.name, True,
                    latency_ms=int((time.monotonic() - collect_start) * 1000),
                    articles_fetched=len(articles),
                    articles_stored=stored,
                )
                logger.info("[Collect] Processed | source=%s | stored=%d", source.id, stored)

            except Exception as e:
                logger.error("[Collect] Source failed | %s | %s", source.id, e)
                # 采集失败指标（Errors）
                await record_collect_run(
                    source.id, source.name, False,
                    latency_ms=int((time.monotonic() - collect_start) * 1000),
                    articles_fetched=0, articles_stored=0,
                    error=str(e)[:500],
                )

        logger.info("[Collect] Manual collection complete | keyword=%r", keyword)

    except Exception as e:
        logger.error("[Collect] Collection failed | %s", e)


@news_router.post("/collect")
async def trigger_collect(req: CollectRequest):
    """手动触发新闻采集任务（异步执行，立即返回）"""
    if req.priority not in ("high", "normal", "low"):
        raise HTTPException(status_code=400, detail="priority must be high/normal/low")

    asyncio.create_task(_run_collection(req.keyword, req.priority))
    return {
        "status": "accepted",
        "message": f"采集任务已启动（priority={req.priority}"
                   + (f", keyword={req.keyword}" if req.keyword else "")
                   + "）",
    }


# ──────────────────────────────────────────────
# Sources
# ──────────────────────────────────────────────

@news_router.get("/sources")
async def list_sources(
    enabled_only: bool = Query(True, description="仅返回启用的源"),
):
    """列出新闻源"""
    sources = await news_storage.list_sources(enabled_only=enabled_only)
    return {
        "sources": [_serialize(s) for s in sources],
        "total": len(sources),
    }


@news_router.get("/sources/health")
async def get_sources_health(
    days: int = Query(30, ge=1, le=365, description="指标统计时间范围（天）"),
):
    """Source Health 面板数据

    按源展示 Latency/Errors/Articles/Duplicates 四项指标 + 覆盖率；
    异常源（最近失败或连续失败）红色标记。
    """
    result = await news_storage.get_source_health(days=days)
    result["sources"] = [_serialize(s) for s in result["sources"]]
    return result


class SourceEnabledRequest(BaseModel):
    enabled: bool = Field(..., description="是否启用该源")


@news_router.post("/sources/{source_id}/enabled")
async def set_source_enabled(source_id: str, req: SourceEnabledRequest):
    """启停一个新闻源（NIC-C3）

    UI 停用后下一采集周期不再采集该源（持久化到 news_sources.yaml）。
    """
    from collectors.source_registry import source_registry

    ok = source_registry.set_enabled(source_id, req.enabled)
    if not ok:
        raise HTTPException(status_code=404, detail=f"Source '{source_id}' not found")
    return {
        "source_id": source_id,
        "enabled": req.enabled,
        "status": "ok",
        "message": "源已" + ("启用" if req.enabled else "停用"),
    }
